src/util/sighash.rs

```diff
@@ -0,0 +1,836 @@
+        // if no annex is present, or 1 otherwise
+        let mut spend_type = 0u8;
+        if annex.is_some() {
+            spend_type |= 1u8;
+        }
+        if script_path.is_some() {
+            spend_type |= 2u8;
+        }
+        spend_type.consensus_encode(&mut writer)?;
+
+        // If hash_type & 0x80 equals SIGHASH_ANYONECANPAY:
+        //      outpoint (36): the COutPoint of this input (32-byte hash + 4-byte little-endian).
+        //      amount (8): value of the previous output spent by this input.
+        //      scriptPubKey (35): scriptPubKey of the previous output spent by this input, serialized as script inside CTxOut. Its size is always 35 bytes.
+        //      nSequence (4): nSequence of this input.
+        if anyone_can_pay {
+            let txin = &self.tx.input[input_index];
+            let previous_output = prevouts.get(input_index)?;
+            txin.previous_output.consensus_encode(&mut writer)?;
+            previous_output.value.consensus_encode(&mut writer)?;
+            previous_output
+                .script_pubkey
+                .consensus_encode(&mut writer)?;
+            txin.sequence.consensus_encode(&mut writer)?;
+        } else {
+            (input_index as u32).consensus_encode(&mut writer)?;
+        }
+
+        // If an annex is present (the lowest bit of spend_type is set):
+        //      sha_annex (32): the SHA256 of (compact_size(size of annex) || annex), where annex
+        //      includes the mandatory 0x50 prefix.
+        if let Some(annex) = annex {
+            let mut enc = sha256::Hash::engine();
+            annex.as_bytes().to_vec().consensus_encode(&mut enc)?;
+            let hash = sha256::Hash::from_engine(enc);
+            hash.consensus_encode(&mut writer)?;
+        }
+
+        // * Data about this output:
+        // If hash_type & 3 equals SIGHASH_SINGLE:
+        //      sha_single_output (32): the SHA256 of the corresponding output in CTxOut format.
+        if sighash == SigHashType::Single {
+            let mut enc = sha256::Hash::engine();
+            self.tx.output[input_index].consensus_encode(&mut enc)?;
+            let hash = sha256::Hash::from_engine(enc);
+            hash.consensus_encode(&mut writer)?;
+        }
+
+        //     if (scriptpath):
+        //         ss += TaggedHash("TapLeaf", bytes([leaf_ver]) + ser_string(script))
+        //         ss += bytes([0])
+        //         ss += struct.pack("<i", codeseparator_pos)
+        if let Some(ScriptPath {
+            script,
+            leaf_version,
+            code_separator_pos,
+        }) = script_path
+        {
+            let mut enc = TapLeafHash::engine();
+            leaf_version.consensus_encode(&mut enc)?;
+            script.consensus_encode(&mut enc)?;
+            let hash = TapLeafHash::from_engine(enc);
+
+            hash.into_inner().consensus_encode(&mut writer)?;
+            0u8.consensus_encode(&mut writer)?;
```

Comment:
Maybe make `KEY_VERSION_0` a constant or add a comment? It took me a while to find what its purpose is.

Reply:
added in 95fd41c